### PR TITLE
disttask: add on subtask finished interface for scheduler

### LINKED ISSUE #43153: add OnSubtaskFinished interface

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: close #43153 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: disttask/framework/scheduler/scheduler.go [hunk 1]
```diff
@@ -162,6 +162,15 @@ func (s *InternalSchedulerImpl) Run(ctx context.Context, task *proto.Task) error
 			}
 			break
 		}
+		if err := scheduler.OnSubtaskFinished(runCtx, subtask.Meta); err != nil {
```
Comment (zimulala) [inline disttask/framework/scheduler/scheduler.go:165]: Could we put this  logic before line157, then we can remove line167-173

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**